In [1]:
! git clone https://github.com/ChandlerBang/GCond.git
%cd GCond

Cloning into 'GCond'...
remote: Enumerating objects: 374, done.
remote: Counting objects: 100% (151/151), done.
remote: Compressing objects: 100% (117/117), done.
remote: Total 374 (delta 70), reused 70 (delta 25), pack-reused 223 (from 1)
Receiving objects: 100% (374/374), 77.44 MiB | 13.45 MiB/s, done.
Resolving deltas: 100% (87/87), done.
/content/GCond


In [5]:
# 看看 cuda version
!nvidia-smi

Fri Dec  5 18:20:41 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   38C    P8              9W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [7]:
# 看看 python 版本
import sys
print(sys.version)

3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]


In [8]:
# 移除所有可能衝突的舊版本
!pip uninstall -y torch torchvision torchaudio torch-scatter torch-sparse torch-cluster torch-spline-conv torch-geometric pyg_lib

# 安裝 PyTorch 2.4.0 (指定 CUDA 12.4 版本)，確保主程式與您的系統環境 (CUDA 12.4) 完美契合。
!pip install torch==2.4.0 torchvision==0.19.0 torchaudio==2.4.0 --index-url https://download.pytorch.org/whl/cu124

# 安裝 PyG 的擴充套件 (指定對應 torch-2.4.0 + cu124 的版本)
!pip install pyg_lib torch_scatter torch_sparse torch_cluster torch_spline_conv -f https://data.pyg.org/whl/torch-2.4.0+cu124.html

# 安裝 PyTorch Geometric 主程式
!pip install torch_geometric

# 安裝其他實驗需要的套件
!pip install numpy scipy tqdm ogb scikit_learn deeprobust

Found existing installation: torch 2.9.0+cu126
Uninstalling torch-2.9.0+cu126:
  Successfully uninstalled torch-2.9.0+cu126
Found existing installation: torchvision 0.24.0+cu126
Uninstalling torchvision-0.24.0+cu126:
  Successfully uninstalled torchvision-0.24.0+cu126
Found existing installation: torchaudio 2.9.0+cu126
Uninstalling torchaudio-2.9.0+cu126:
  Successfully uninstalled torchaudio-2.9.0+cu126
Found existing installation: torch_scatter 2.1.2+pt24cu118
Uninstalling torch_scatter-2.1.2+pt24cu118:
  Successfully uninstalled torch_scatter-2.1.2+pt24cu118
Found existing installation: torch_sparse 0.6.18+pt24cu118
Uninstalling torch_sparse-0.6.18+pt24cu118:
  Successfully uninstalled torch_sparse-0.6.18+pt24cu118
Found existing installation: torch_cluster 1.6.3+pt24cu118
Uninstalling torch_cluster-1.6.3+pt24cu118:
  Successfully uninstalled torch_cluster-1.6.3+pt24cu118
Found existing installation: torch_spline_conv 1.2.2+pt24cu118
Uninstalling torch_spline_conv-1.2.2+pt24cu118:
 

In [9]:
# 確認一下版本都沒問題!!
import torch
print(f"PyTorch version: {torch.__version__}")
print(f"CUDA version: {torch.version.cuda}")
print(f"CUDA available: {torch.cuda.is_available()}")

# 測試能不能成功 import 這些容易報錯的套件
import torch_sparse
import torch_scatter
print("PyG extensions imported successfully!")

PyTorch version: 2.4.0+cu124
CUDA version: 12.4
CUDA available: True
PyG extensions imported successfully!


In [2]:
"""
# 安裝PyTorch
! pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118
# 安裝PyTorch Geometric
! pip install pyg_lib torch_scatter torch_sparse torch_cluster torch_spline_conv \
    -f https://data.pyg.org/whl/torch-2.4.0+cu118.html
! pip install torch_geometric
# 其他依賴
! pip install numpy scipy tqdm ogb scikit_learn deeprobust


# 2. 強制安裝指定的 PyTorch 2.4.0 版本 (搭配 CUDA 11.8)
!pip install torch==2.4.0 torchvision==0.19.0 torchaudio==2.4.0 --index-url https://download.pytorch.org/whl/cu118

# 3. 安裝對應 PyTorch 2.4.0 + CUDA 11.8 的 PyG 擴充套件
!pip install pyg_lib torch_scatter torch_sparse torch_cluster torch_spline_conv -f https://data.pyg.org/whl/torch-2.4.0+cu118.html

# 4. 最後安裝 PyTorch Geometric 與其他套件
!pip install torch_geometric
!pip install numpy scipy tqdm ogb scikit_learn deeprobust
"""


Looking in indexes: https://download.pytorch.org/whl/cu118
Looking in links: https://data.pyg.org/whl/torch-2.4.0+cu118.html
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 88.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.3/10.3 MB 116.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.0/5.0 MB 111.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 100.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 945.6/945.6 kB 60.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.7/63.7 kB 5.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 78.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.8/78.8 kB 6.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 219.4/219.4 kB 19.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.9/27.9 MB 49.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 87.2/87.2 kB 8.9 MB/s eta 0:00:00


In [10]:
! python train_gcond_transduct.py --dataset cora --nlayers=2 --lr_feat=1e-4 --gpu_id=0 --lr_adj=1e-4 --r=0.5 --epochs=400


Namespace(gpu_id=0, dataset='cora', dis_metric='ours', epochs=400, nlayers=2, hidden=256, lr_adj=0.0001, lr_feat=0.0001, lr_model=0.01, weight_decay=0.0, dropout=0.0, normalize_features=True, keep_ratio=1.0, reduction_rate=0.5, seed=15, alpha=0, debug=0, sgc=1, inner=0, outer=20, save=0, one_step=0)
Processing...
Done!
size of adj_train: (140, 140)
#edges in adj_train: 42.0
adj_syn: (70, 70) feat_syn: torch.Size([70, 1433])
/usr/local/lib/python3.12/dist-packages/deeprobust/graph/utils.py:356: UserWarning: torch.sparse.SparseTensor(indices, values, shape, *, device=) is deprecated.  Please use torch.sparse_coo_tensor(indices, values, shape, dtype=, device=). (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:641.)
  return torch.sparse.FloatTensor(sparseconcat.t(),sparsedata,torch.Size(sparse_mx.shape))
/content/GCond/utils.py:177: UserWarning: 'data.NeighborSampler' is deprecated, use 'loader.NeighborSampler' instead
  self.samplers.append(NeighborSampler(adj,
/usr/local/lib/

TODO 1:

size of adj_train: (140, 140)代表訓練集中140個訓練節點的adjacency matrix，size of adj_syn: (70, 70)代表生成圖的adjacency matrix，請問這樣我的生成的凝結圖剩下多少節點?

**ANS: 生成的凝結圖（Condensed Graph）剩下 70 個節點。**   

size of adj_train: (140, 140)          
edges in adj_train: 42.0            
adj_syn: (70, 70) feat_syn: torch. Size([70, 1433])   


推論原因：

1.   從 Log 觀察： 程式執行結果顯示 adj_syn: (70, 70)。在圖學中，鄰接矩陣（Adjacency Matrix）的維度為 $(N \times N)$，其中 $N$ 代表節點數量。因此 $(70, 70)$ 的矩陣意味著圖中有 70 個節點。
2.   從參數驗證： 本次實驗設定的壓縮率（Reduction Rate, r）為 0.5（指令中的 --r=0.5）。原始訓練圖有 140 個節點，計算 $140 \times 0.5 = 70$，與生成的矩陣大小相符。


TODO 2:

上面的範例是論文作者使用一種GCN模型來訓練CORA資料集，然後透過gradient matching的方法(這裡用的是GCond)製作一個合成圖，之後我使用一個自製的GCN模型對合成圖進行測試但發現準確度和論文的方法相比奇差無比(剩下30%左右)，請簡述一下可能的原因。

(Hint:使用GCond方法合成圖時，是不是會有模型綁定的問題?)



**ANS: 模型耦合 (Model Coupling/Binding) 或 對特定架構的過擬合。**  


GCond 的核心機制是 Gradient Matching，優化目標是讓「合成圖在特定模型產生的梯度」去逼近「原始圖在該模型產生的梯度」，因此生成的合成圖 $S$ 實際是為了欺騙那個特定的 GCN 架構而生成的，它將資訊編碼成了該特定模型容易讀取的特徵分佈。

當您換了一個自製的 GCN 模型，在層數、激活函數或聚合方式都可能不同的情況下，模型在參數空間中的梯度方向會完全不同，原本合成圖中隱藏的特徵資訊對新模型來說可能變成了雜訊或無法解讀的數據，導致準確率大幅下降。